zingolib/src/wallet/transaction_records_by_id.rs

```diff
@@ -376,6 +365,51 @@ impl TransactionRecordsById {
 
         self.invalidate_transactions(txids_to_remove);
     }
+
+    /// Note this method is INCORRECT in the case of a 0-value, 0-fee transaction from the
+    /// Creating Capability.  Such a transaction would violate ZIP317, but could exist in
+    /// the Zcash protocol
+    ///  TODO:   Test and handle 0-value, 0-fee transaction
+    pub(crate) fn transaction_kind(&self, query_record: &TransactionRecord) -> TransactionKind {
+        let mut sapling_spends: Vec<&SaplingNote> =
+            Vec::with_capacity(query_record.spent_sapling_nullifiers.len());
+        query_record
+            .spent_sapling_nullifiers()
+            .iter()
+            .for_each(|nullifier| {
+                if let Some(spend) = self.find_sapling_spend(nullifier) {
+                    sapling_spends.push(spend);
+                }
+            });
+        let mut orchard_spends: Vec<&OrchardNote> =
+            Vec::with_capacity(query_record.spent_orchard_nullifiers.len());
+        query_record
+            .spent_orchard_nullifiers()
+            .iter()
+            .for_each(|nullifier| {
+                if let Some(spend) = self.find_orchard_spend(nullifier) {
+                    orchard_spends.push(spend);
+                }
+            });
+
+        if sapling_spends.is_empty()
+            && orchard_spends.is_empty()
+            && query_record.total_transparent_value_spent == 0
+            && query_record.outgoing_tx_data.is_empty()
+        {
+            TransactionKind::Received
+        } else if sapling_spends.is_empty()
+            && orchard_spends.is_empty()
+            && query_record.total_transparent_value_spent > 0
+            && query_record.outgoing_tx_data.is_empty()
+        {
+            // TODO: this could be improved by checking outputs recipient addr against the wallet addrs
+            TransactionKind::Sent(SendType::Shield)
```

Comment:
Won't this no longer be true in ZIP320?

Reply:
this still holds for zip320 because when sending to exchange with 2-step the second step will still create outgoing tx data OR the address will not match our own if we change the logic as the TODO describes

Reply:
we will likely add TEX as a SendType variant

Reply:
ok for now